File: saucybot/sites/newgrounds.py
import requests
import io
import discord
import json
from pyquery import PyQuery as pq
from sites.base import Base


class Newgrounds(Base):

    def __init__(self):
        self.name = 'Newgrounds'
        self.base_url = 'https://www.newgrounds.com'
        self.post_url = 'https://www.newgrounds.com/art/view/{}/{}'
        self.pattern = 'newgrounds.com\/art\/view\/(.*)\/(.*)'
        self.colour = discord.Colour(0xFFF17A)

    def process(self, match):
        (ng_user, ng_slug) = match.groups()

        session = requests.Session()

        full_url = self.post_url.format(ng_user, ng_slug)

        response = session.get(full_url)

        d = pq(response.text)

        title = d('.body-guts .column.wide.right .pod-head h2')
        image = d('#portal_item_view img')
        author_link = d('.body-guts .column.thin .pod-body .item-details a')
        author_img = d(
            '.body-guts .column.thin .pod-body .user-icon-bordered img')
        description = d('#author_comments')
        score = d('#score_number')

        ret = {}

        description_text = description.text()

        if len(description_text) > 300:
            description_text = description_text[0:300] + '...'

        discord_embed = discord.Embed(
            title=title.text(), url=full_url, description=description_text, colour=self.colour)

        discord_embed.set_image(url=image.attr('src'))

        discord_embed.set_author(name=author_link.text(
        ), url='https:' + author_link.attr('href'), icon_url='https:' + author_img.attr('src'))

        discord_embed.add_field(name='Score', value='{} / 5.00'.format(score.text()))

        ret['embed'] = discord_embed

        return ret

File: src/sites/Newgrounds.ts
import BaseSite from './BaseSite';
import ProcessResponse from './ProcessResponse';
import cheerio from 'cheerio';
import got from 'got';
import { Message, EmbedBuilder } from 'discord.js';
import CacheManager from '../CacheManager';
import { processDescription } from '../Helpers';

class Newgrounds extends BaseSite {
    identifier = 'Newgrounds';

    pattern =
        /https?:\/\/(www\.)?newgrounds\.com\/art\/view\/(?<user>.*)\/(?<slug>\S+)\/?/gim;

    color = 0xfff17a;

    async process(
        match: RegExpMatchArray,
        /* eslint-disable @typescript-eslint/no-unused-vars */
        source: Message | null
    ): Promise<ProcessResponse | false> {
        const message: ProcessResponse = {
            embeds: [],
            files: [],
        };

        if (!match.groups?.user || !match.groups?.slug) {
            return Promise.resolve(false);
        }

        const url = match[0];

        const body = await this.getPage(
            url,
            match.groups.user,
            match.groups.slug
        );

        if (!body) {
            return Promise.resolve(false);
        }

        const $ = cheerio.load(body);

        const title = $('.body-guts .column.wide.right .pod-head h2');
        const image = $('.pod-body .image #portal_item_view img');

        const description = $('#author_comments');

        const authorLink = $(
            '.body-guts .column.thin .pod-body .item-details a'
        );
        const authorImage = $(
            '.body-guts .column.thin .pod-body .user-icon-bordered img'
        );

        const views = $('.sidestats dt:contains("Views")').siblings().first();

        const score = $('#score_number');

        const embed = new EmbedBuilder({
            title: title.text(),
            url: url,
            description: processDescription(description?.html() ?? ''),
            color: this.color,
            image: {
                url: image.attr('src') ?? '',
            },
            author: {
                name: authorLink.text(),
                url: authorLink.attr('href'),
                iconURL: authorImage.attr('src'),
            },
            fields: [
                {
                    name: 'Views',
                    value: views.text(),
                    inline: true,
                },
                {
                    name: 'Score',
                    value: `${score.text()} / 5.00`,
                    inline: true,
                },
            ],
            footer: {
                text: 'Newgrounds',
            },
        });

        message.embeds.push(embed);

        return Promise.resolve(message);
    }

    async getPage(
        url: string,
        user: string,
        slug: string
    ): Promise<string | null> {
        const cacheKey = `newgrounds.art_${user}_${slug}`;
        const cacheManager = await CacheManager.getInstance();

        const cachedValue = await cacheManager.remember(cacheKey, async () => {
            const response = await got.get(url);
            return Promise.resolve(response.body);
        });

        if (!cachedValue) {
            return Promise.resolve(null);
        }

        return Promise.resolve(cachedValue);
    }
}

export default Newgrounds;
